compiler/rustc_driver_impl/src/lib.rs

```diff
@@ -1125,19 +1119,31 @@ pub fn describe_flag_categories(early_dcx: &EarlyDiagCtxt, matches: &Matches) ->
     }
 
     if cg_flags.iter().any(|x| *x == "passes=list") {
-        let backend_name = debug_flags.iter().find_map(|x| x.strip_prefix("codegen-backend="));
-
-        let opts = config::Options::default();
-        let sysroot = filesearch::materialize_sysroot(opts.maybe_sysroot.clone());
-        let target = config::build_target_config(early_dcx, &opts, &sysroot);
-
-        get_codegen_backend(early_dcx, &sysroot, backend_name, &target).print_passes();
+        get_backend_from_raw_matches(early_dcx, matches).print_passes();
         return true;
     }
 
     false
 }
 
+/// Get the codegen backend based on the raw [`Matches`].
+///
+/// `rustc -vV` and `rustc -Cpasses=list` need to get the codegen backend before we have parsed all
+/// arguments and creates a [`Session`]. This function reads `-Zcodegen-backend` and `--sysroot`
+/// without validating any other arguments and loads the codegen backend based on these arguments.
+fn get_backend_from_raw_matches(
+    early_dcx: &EarlyDiagCtxt,
+    matches: &Matches,
+) -> Box<dyn CodegenBackend> {
+    let debug_flags = matches.opt_strs("Z");
+    let backend_name = debug_flags.iter().find_map(|x| x.strip_prefix("codegen-backend="));
+    let opts = config::Options::default();
+    let sysroot = filesearch::materialize_sysroot(matches.opt_str("sysroot").map(PathBuf::from));
+    let target = config::build_target_config(early_dcx, &opts, &sysroot);
```

Comment:
Why build_target_config uses opts that is config::Options::default(), which sets target_triple to `host_tuple`? This looks weird for function named `build_target_config` use some defaulted target.

Reply:
I've thought about inlining `build_target_config` and directly using the host tuple instead, but the function is just a bit too complicated to inline IMO.

Reply:
I mean, that this will potentially ignore actual `target_triple` and use `host_tuple()` instead when calling `get_codegen_backend`?

Reply:
Why would the backend version depend on the target. Same for the pass list I assume, whether they are enabled may be dynamic if there were target-dependent passes.

Reply:
For version - yes, but for passes - idk, i'll try to check.

Reply:
The target can affect the codegen backend we use through the `default_codegen_backend` field in the target spec. We're currently ignoring this already however.

Reply:
Okay, can you add comment that using default target in this function is ok?

Reply:
It is technically not correct, but it is not a regression from the current state and using `--target` could potentially break things.

Reply:
I've added `--target` support too.